### ISSUE TITLE
Support N-dimensional data arrays as input in the request data array

### ISSUE THREAD

**Issue by pvaneck** [kserve/rest-proxy#5]
Currently, the rest-proxy expects the input data array to be 1-D. However the KServe Predict v2 REST API [allows for passing in data in a multi-dimensional format](https://github.com/kserve/kserve/blob/master/docs/predict-api/v2/required_api.md#parameters).

With gRPC, however, the data must be the flattened, one-dimensional, row-major order of the tensor elements.

Some logic in the proxy is needed to flatten an arbitrary multi-dimensional array.

Example:

The following works since the array is flattened and 1-D.
```shell
curl -X POST -k http://localhost:8080/v2/models/example-mnist-predictor/infer -d '{ "id": "foo", "inputs": [{ "name": "predict", "shape": [2, 64], "datatype": "FP32", "data": [0.0, 0.0, 1.0, 11.0, 14.0, 15.0, 3.0, 0.0, 0.0, 1.0, 13.0, 16.0, 12.0, 16.0, 8.0, 0.0, 0.0, 8.0, 16.0, 4.0, 6.0, 16.0, 5.0, 0.0, 0.0, 5.0, 15.0, 11.0, 13.0, 14.0, 0.0, 0.0, 0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0, 0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0, 0.0, 0.0, 1.0, 11.0, 14.0, 15.0, 3.0, 0.0, 0.0, 1.0, 13.0, 16.0, 12.0, 16.0, 8.0, 0.0, 0.0, 8.0, 16.0, 4.0, 6.0, 16.0, 5.0, 0.0, 0.0, 5.0, 15.0, 11.0, 13.0, 14.0, 0.0, 0.0, 0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0, 0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0]}]}'
```

This doesn't yet work, but should:
```shell
curl -X POST -k http://localhost:8080/v2/models/example-mnist-predictor/infer -d '{ "id": "foo", "inputs": [{ "name": "predict", "shape": [2, 64], "datatype": "FP32", "data": [[0.0, 0.0, 1.0, 11.0, 14.0, 15.0, 3.0, 0.0, 0.0, 1.0, 13.0, 16.0, 12.0, 16.0, 8.0, 0.0, 0.0, 8.0, 16.0, 4.0, 6.0, 16.0, 5.0, 0.0, 0.0, 5.0, 15.0, 11.0, 13.0, 14.0, 0.0, 0.0, 0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0, 0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0],[0.0, 0.0, 1.0, 11.0, 14.0, 15.0, 3.0, 0.0, 0.0, 1.0, 13.0, 16.0, 12.0, 16.0, 8.0, 0.0, 0.0, 8.0, 16.0, 4.0, 6.0, 16.0, 5.0, 0.0, 0.0, 5.0, 15.0, 11.0, 13.0, 14.0, 0.0, 0.0, 0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0, 0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0]]}]}'
```

**Comment by ScrapCodes** [comment]
Hi @pvaneck, It is becoming difficult to generalise it to Nth dim, do you think we can support a reasonable value for N?

The reason for the difficulty is, underlying json gets parsed into a single dimension []interface array. And the type of each item is, interface{} regardless of the fact that input array was a 2D/3D etc..

**Comment by pvaneck** [comment]
@ScrapCodes I think that is fine if it will help simplify the implementation. What do you think is reasonably doable? I'm thinking at least up to 4-D to handle batches of images.

**Comment by ScrapCodes** [comment]
Sounds good! @pvaneck, do you know any example models which I can use for testing and support 3D and 4D inputs? `example-mnist-predictor` supports only 2D inputs.

**Comment by pvaneck** [comment]
Hmm, perhaps we can try testing with some simple image models, like a MobileNet model in PyTorch or TensorFlow.
